### ISSUE TITLE
Speed is much slower than with modified (not by me) BinomialLLC 2018 version

### ISSUE THREAD

**Issue by suphamster** [DaemonEngine/crunch#26]
I've decided to test compressing speed (with default settings in both cases) and it was at least 3 times slower. Compiled latest version with VS2017 x64 build. I've attached both version (that I've compiled and old one) so anybody can test it by yourself.
[crunch.zip](https://github.com/DaemonEngine/crunch/files/8879281/crunch.zip)
P.S. It seems that problem happened coz I was using debug version. Release version has same speed like modified BinomialLLC 2018.

**Comment by illwieckz** [comment]
I'm surprised you get the same compression time. Having the newer one being slower as obviously wrong (Debug build explains that), but not having the newer one faster is very surprising as well. To [quote myself](https://unvanquished.net/unvanquished-area-51/):

> [Unity guys said](https://blog.unity.com/technology/crunch-compression-of-etc-textures) that their modified crunch tool “can compress up to 2.5 times faster, while providing about 10% better compression ratio”. So We did a test on our [own asset repository](https://github.com/UnvanquishedAssets/UnvanquishedAssets), re-crunching all the resources and textures packages. At the time we did the the test the given corpus produced 1797 .crn files. We benchmarked both our old crunch tool based on BinomialLLC code and the new Unity one with our patches applied. We rebuilt the binaries so there is no bias on that part. And here come the results :
> 
> BinomialLLC’s crunch:
> 
>     compression time: 115 minutes 8 seconds
>     produced file size: 269 megabytes
> 
> Unity’s crunch:
> 
>     compression time: 26 minutes 43 seconds
>     produced file size: 239 megabytes

This is basically why we switched to the Unity variant, it both saves time and size in obvious ways. Though, maybe the difference is not that seen when done on a couple of specific files, but on large corpus this is impressive.

**Comment by suphamster** [comment]
> I'm surprised you get the same compression time. Having the newer one being slower as obviously wrong (Debug build explains that), but not having the newer one faster is very surprising as well. To [quote myself](https://unvanquished.net/unvanquished-area-51/):
> 
> > [Unity guys said](https://blog.unity.com/technology/crunch-compression-of-etc-textures) that their modified crunch tool “can compress up to 2.5 times faster, while providing about 10% better compression ratio”. So We did a test on our [own asset repository](https://github.com/UnvanquishedAssets/UnvanquishedAssets), re-crunching all the resources and textures packages. At the time we did the the test the given corpus produced 1797 .crn files. We benchmarked both our old crunch tool based on BinomialLLC code and the new Unity one with our patches applied. We rebuilt the binaries so there is no bias on that part. And here come the results :
> > BinomialLLC’s crunch:
> > ```
> > compression time: 115 minutes 8 seconds
> > produced file size: 269 megabytes
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Unity’s crunch:
> > ```
> > compression time: 26 minutes 43 seconds
> > produced file size: 239 megabytes
> > ```
> 
> This is basically why we switched to the Unity variant, it both saves time and size in obvious ways. Though, maybe the difference is not that seen when done on a couple of specific files, but on large corpus this is impressive.
> 
> _Edit:_ my comment at the time said “our old crunch tool” but I also tested latest upstream from Binomial and it was as slow.

I've tested on my own 1061 dds files collection and it was just about 15 sec faster (total time about 10 min). Btw in both cases CPU usage was never more than 60% (i7-8700).

**Comment by illwieckz** [comment]
What is your compression scenario, i.e. you compress from what to what?

In the tests I've quoted we were compressing from png to crn because for our game we have lossless image source for almost every texture.

If you're converting dds to crn I have no idea how it performs, and I would not be surprised if crunch would skip some computations when converting from dds to crn, assuming some computations are already done because crn is somewhat like a custom dds container. If yes the compute time difference may not be that huge, but that's just an hypothesis (I haven't checked).

**Comment by suphamster** [comment]
Yes, I'm converting only dds to crn. Btw is it possible to utilize more CPU resources (90% at least) in ur tool?

**Comment by illwieckz** [comment]
I don't think this tool will use more CPU than others for the same task. If you want to put your CPU on knees you may try those various techniques:

First, your CPU having 8 cores, 12 threads, you may try to force crunch to use more threads with `-helperThreads N` (up to 15, the default being the amount of core minus one, clamped to 15 if greater).

And what I do myself is that I use tools like `xargs` or GNU `parallel` to run multiple instances of `crunch` on multiple images at the same time.

For example, if you have a bash shell (provided with Git if I'm right), given the command is called within a folder full of dds files, you may be able to do:

```sh
find . -type f -name '*.dds' -print0 | xargs -0 -I{} -P$(nproc) crunch {} -fileformat crn
```

This would look for every dds file in the current directory and spawn one crunch conversion task per cpu core/thread until everything is done.

If you have GNU parallel you can do the same but with a progress bar:

```sh
find . -type f -name '*.dds' -print0 | parallel --will-cite --bar -0 -I{} -P$(nproc) crunch {} -fileformat crn
```

With `$(nproc)` the amount of CPU cores. Since Crunch already uses up to 16 cores, so spawning 16 crunches spawning 16 cores will put the CPU on knees, you can also reduce amount of cores used per `crunch` call with `-helperThreads`.

For example you can do:

```sh
find . -type f -name '*.dds' -print0 | xargs -0 -I{} -P4 crunch {} -fileformat crn -helperThreads 6
```

This will compress 4 images in parallel, every task using 6 cores, so basically the 12 threads cpu will run 24 threads, probably enough to reach 100%.
